### PR TITLE
[Icons] Prepare lock / warmup optimization

### PR DESCRIPTION
**Pull request by smnandre**
Method needed for both
* lock/warmup optimization
* sprite generation
* search performances

### CODE REVIEW

File: src/Icons/src/Iconify.php [hunk 1]
```diff
@@ -85,6 +85,52 @@ public function fetchIcon(string $prefix, string $name): Icon
...
+            'query' => [
+                'icons' => $queryString,
+            ],
+        ]);
+
+        if (200 !== $response->getStatusCode()) {
+            throw new IconNotFoundException(\sprintf('The icon set "%s" does not exist on iconify.design.', $prefix));
+        }
+
+        $data = $response->toArray();
```
Comment (Kocal) [inline src/Icons/src/Iconify.php]: Can't we support more than 500 chars by splitting icons and doing multiple requests?

File: src/Icons/src/Iconify.php [hunk 2]
```diff
@@ -85,6 +85,52 @@ public function fetchIcon(string $prefix, string $name): Icon
...
+        // Sort to enforce cache hits
+        $queryString = implode(',', $names);
+        if (!preg_match('#^[\w-]+(,[\w-]+)*$#', $queryString)) {
+            throw new \InvalidArgumentException('Invalid icon names.');
+        }
+
+        // 500 chars max
+        // 42 chars: https://api.iconify.design/mdi.json?icons=
+        if (450 < \strlen($queryString)) {
+            throw new \InvalidArgumentException('The query string is too long.');
```
Comment (Kocal) [inline src/Icons/src/Iconify.php]: Please use dedicated exception classes

File: src/Icons/src/Iconify.php (hunk 1 above)
Comment (Kocal) [reply src/Icons/src/Iconify.php]: Ah, didn't see #2352

**Review by Kocal — APPROVED** [review]

File: src/Icons/src/Iconify.php (hunk 2 above)
Comment (kbond) [reply src/Icons/src/Iconify.php]: Is the plan to catch this? (I typically only add named exceptions if this is the case - or it's realistic an end-user would want to catch)

File: src/Icons/src/Iconify.php [hunk 3]
```diff
@@ -85,6 +85,52 @@ public function fetchIcon(string $prefix, string $name): Icon
...
     }
 
+    public function fetchIcons(string $prefix, array $names): array
+    {
+        if (!isset($this->sets()[$prefix])) {
+            throw new IconNotFoundException(\sprintf('The icon set "%s" does not exist on iconify.design.', $prefix));
+        }
+
+        // Sort to enforce cache hits
+        $queryString = implode(',', $names);
```
Comment (javiereguiluz) [inline src/Icons/src/Iconify.php:96]: the comment above says "sort" (to get more cache hits) but here you are not sorting the values, just imploding them. The sorting of the `$names` array is missing, right?

File: src/Icons/src/Iconify.php (hunk 2 above)
Comment (Kocal) [reply src/Icons/src/Iconify.php]: Nah I don't think so, but I like to have dedicated exceptions and also a base exception class per package (like they do on Symfony components)

Comment (smnandre) [reply src/Icons/src/Iconify.php]: Not on every component to be far... ;) 

Yet I kinda agree to use custom exception but this should be done in a dedicated PR i think, carefully selecting the one we want to "migrate" and the one we do not. 

To explain myself here, this is internal and "should never" happen. 

But as we all know what "should never" means... i get your point.

**Review by kbond — APPROVED** [review]

**Pull request merged**